lib/exp.js

```diff
@@ -0,0 +1,1206 @@
+
+const _cmpExp = (op) => (left, right) => [{ op, count: 3 }].concat(left, right)
+
+exports.eq = _cmpExp(exp.ops.CMP_EQ)
+
+exports.ne = _cmpExp(exp.ops.CMP_NE)
+const _ne = exports.ne
+
+exports.gt = _cmpExp(exp.ops.CMP_GT)
+
+exports.ge = _cmpExp(exp.ops.CMP_GE)
+
+exports.lt = _cmpExp(exp.ops.CMP_LT)
+
+exports.le = _cmpExp(exp.ops.CMP_LE)
+
+exports.cmpRegex = (options, regex, cmpStr) => [
+  { op: exp.ops.CMP_REGEX, count: 4 },
+  ..._int(options),
+  ..._rawStr(regex),
+  ...cmpStr
+]
+
+exports.cmpGeo = _cmpExp(exp.ops.CMP_GEO)
+
+/*********************************************************************************
+ * LOGICAL EXPRESSIONS
+ *********************************************************************************/
+
+exports.not = (expr) => [
+  { op: exp.ops.NOT, count: 2 },
+  ...expr
+]
+
+const _VAExp = (op) => (...expr) => [].concat(
+  { op },
+  ...expr,
+  { op: exp.ops.END_OF_VA_ARGS }
+)
+
+exports.and = _VAExp(exp.ops.AND)
+
+exports.or = _VAExp(exp.ops.OR)
+
+/*********************************************************************************
+ * ARITHMETIC EXPRESSIONS
+ *********************************************************************************/
+
+const _VAargs = (op) => (...expr) => [].concat(
+  { op },
+  ...expr,
+  { op: exp.ops.END_OF_VA_ARGS }
+)
+
+exports.add = _VAargs(exp.ops.ADD)
+exports.sub = _VAargs(exp.ops.SUB)
+exports.mul = _VAargs(exp.ops.MUL)
+exports.div = _VAargs(exp.ops.DIV)
+
+const _p3Exp = (op) => (p1, p2) => [
+  { op: op, count: 3 },
+  ...p1,
+  ...p2
+]
+exports.pow = _p3Exp(exp.ops.POW)
+exports.log = _p3Exp(exp.ops.LOG)
+exports.mod = _p3Exp(exp.ops.MOD)
+
+const _p2Exp = (op) => (p1) => [
+  { op: op, count: 2 },
+  ...p1
+]
+exports.abs = _p2Exp(exp.ops.ABS)
+exports.floor = _p2Exp(exp.ops.FLOOR)
+exports.ceil = _p2Exp(exp.ops.CEIL)
+exports.toInt = _p2Exp(exp.ops.TO_INT)
+exports.toFloat = _p2Exp(exp.ops.TO_FLOAT)
+
+exports.intAnd = _VAargs(exp.ops.INT_AND)
+exports.intOr = _VAargs(exp.ops.INT_OR)
+exports.intXor = _VAargs(exp.ops.INT_XOR)
+
+exports.intNot = _p2Exp(exp.ops.INT_NOT)
+
+exports.intLshift = _p3Exp(exp.ops.INT_LSHIFT)
+exports.intRshift = _p3Exp(exp.ops.INT_RSHIFT)
+exports.intArshift = _p3Exp(exp.ops.INT_ARSHIFT)
+
+exports.intCount = _p2Exp(exp.ops.INT_COUNT)
+
+exports.intLscan = _p3Exp(exp.ops.INT_LSCAN)
+exports.intRscan = _p3Exp(exp.ops.INT_RSCAN)
+
+exports.min = _VAargs(exp.ops.MIN)
+exports.max = _VAargs(exp.ops.MAX)
+
+/*********************************************************************************
+ * FLOW CONTROL AND VARIABLE EXPRESSIONS
+ *********************************************************************************/
+exports.cond = _VAargs(exp.ops.COND)
+exports.let = _VAargs(exp.ops.LET)
+
+/*********************************************************************************
+ * LIST READ EXPRESSIONS
+ *********************************************************************************/
+
+function getContextType (ctx, type) {
+  return type
```

Comment:
Need to replace this with the same logic as the C client's [as_exp_get_ctx_type function](https://github.com/aerospike/aerospike-client-c/blob/master/src/main/aerospike/as_exp.c#L425-L435).

Also need to add unit test covering different list contexts, so that this function gets covered.